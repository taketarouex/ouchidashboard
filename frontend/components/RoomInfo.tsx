import React, { FC, useState } from 'react';
import CircularProgress from '@material-ui/core/CircularProgress'
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import { LineChart, XAxis, YAxis, Line, Tooltip, ResponsiveContainer } from 'recharts'
import useSWR from 'swr'
import dayjs, { Dayjs } from 'dayjs'

type Log = {
  Value: number
  UpdatedAt: Date
}

const fetchLogs: (url: string) => Promise<Log[]> = (url) => fetch(url).then(
  res => {
    if (!res.ok) {
      const error = new Error('An error occurred while fetching the data.')
      throw error
    }
    return res.json()
  }
)

const useRoomLog = ({ roomName, logType, start, end }: { roomName: string, logType: string, start: Dayjs, end: Dayjs }) => {
  const startISO = start.toISOString()
  const endISO = end.toISOString()
  const { data, error } = useSWR(`/api/rooms/${roomName}/logs/${logType}?start=${startISO}&end=${endISO}`, fetchLogs)

  return {
    data: data,
    isLoading: !error && !data,
    isError: error
  }
}

const RoomChart = ({ roomName, logType, start, end }: { roomName: string, logType: string, start: dayjs.Dayjs, end: dayjs.Dayjs }) => {
  if (roomName === "") return <div>select rooms</div>
  const { data, isLoading, isError } = useRoomLog({ roomName, logType, start, end })
  if (isLoading) return <CircularProgress />
  if (isError) return <div>error</div>
  return (
    <ResponsiveContainer width={'90%'} height={400}>
      <LineChart
        data={data}
      >
        <YAxis />
        <XAxis dataKey="UpdatedAt" domain={['dataMin', 'dataMax']} />
        <Line type="monotone" dataKey="Value" stroke="#8884d8" />
        <Tooltip />
      </LineChart>
    </ResponsiveContainer>
  )
}

const fetchRoomNames: (url: string) => Promise<string[]> = (url) => fetch(url).then(
  res => {
    if (!res.ok) {
      const error = new Error('An error occurred while fetching the data.')
      throw error
    }
    return res.json()
  }
)

const useRoomNames = () => {
  const { data, error } = useSWR(`/api/rooms`, fetchRoomNames)
  return {
    data: data,
    isLoading: !error && !data,
    isError: error
  }
}

const RoomSelect = ({ room, handleChange }: { room: string, handleChange: (event: React.ChangeEvent<{ value: unknown }>) => void }) => {
  const { data, isLoading, isError } = useRoomNames();
  if (isLoading) return <CircularProgress />
  if (isError) return <div>error</div>
  return (
    <Select
      labelId="room"
      id="room-select"
      value={room}
      onChange={handleChange}
    >
      {data.map(v =>
        <MenuItem value={v} key={v}>{v}</MenuItem>
      )}
    </Select>
  )
}

export const RoomInfo = ({ logType, start, end }: { logType: string, start: Dayjs, end: Dayjs }) => {
  const [room, setRoom] = useState<string>("")
  const handleChange = (event: React.ChangeEvent<{ value: unknown }>) => {
    setRoom(event.target.value as string)
  }
  return (
    <div>
      <RoomSelect room={room} handleChange={handleChange} />
      <RoomChart roomName={room} logType={logType} start={start} end={end} />
    </div>
  )
}