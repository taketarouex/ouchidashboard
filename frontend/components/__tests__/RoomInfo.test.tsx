import React from 'react';
import { render, waitForElementToBeRemoved, fireEvent, within } from '@testing-library/react'
import { RoomInfo } from '../RoomInfo'
import '@testing-library/jest-dom'
import dayjs from 'dayjs'

import { enableFetchMocks } from 'jest-fetch-mock'

test('RoomInfo', async () => {
  enableFetchMocks()
  fetchMock.mockResponses(
    JSON.stringify(["hoge", "fuga", "bar"]),
    JSON.stringify([
      { value: 0, updatedAt: dayjs(new Date(2020, 7, 31, 0, 0, 0, 0)) },
      { value: 1, updatedAt: dayjs(new Date(2020, 7, 31, 1, 0, 0, 0)) },
      { value: 2, updatedAt: dayjs(new Date(2020, 7, 31, 2, 0, 0, 0)) }
    ])
  )
  const start = dayjs(new Date(2020, 7, 31, 0, 0, 0, 0))
  const end = dayjs(new Date(2020, 7, 31, 10, 0, 0, 0))

  const { container, getByRole, getByText } = render(<RoomInfo logType="test" start={start} end={end} />)
  await waitForElementToBeRemoved(getByRole("progressbar"))
  expect(fetchMock).toBeCalledWith("/api/rooms")
  expect(getByText("select rooms")).toBeInTheDocument()
  fireEvent.mouseDown(getByRole('button'))
  const listbox = within(getByRole('listbox'))
  fireEvent.click(listbox.getByText("hoge"))

  await waitForElementToBeRemoved(getByRole("progressbar"))

  expect(fetchMock).toBeCalledWith(
    `/api/rooms/hoge/logs/test?start=${start.toISOString()}&end=${end.toISOString()}`)
  expect(container).toMatchSnapshot()
})
