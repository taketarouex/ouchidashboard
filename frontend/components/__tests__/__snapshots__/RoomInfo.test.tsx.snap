// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomInfo 1`] = `
<div>
  <div>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline Mui-focused Mui-focused"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="room room-select"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="room-select"
        role="button"
        tabindex="0"
      >
        hoge
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        tabindex="-1"
        value="hoge"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      class="recharts-responsive-container"
      id="undefined"
      style="width: 90%; height: 400px;"
    >
      <div />
    </div>
  </div>
</div>
`;
